docs/rfcs/0002-storing-course-data.md

```diff
@@ -0,0 +1,134 @@
+    later if needed (example: we may want to show completion progress or payment
+    history for a user that took some program courses on edx.org and other
+    program course on our Open edX instance)
+
+1. **`CourseRun.external_url`**
+    
+    We need to be able to link directly to a course on its courseware
+    platform. One option is to store the external id and, in our app logic,
+    generate the link to the course based on that id and the platform it
+    exists on (indicated by `source`). Another option is to simply store
+    the external url alongside the external id. That involves some redundant
+    data (the URL will almost certainly contain the external id), but the 
+    external id is highly unlikely to change and we will gain some 
+    flexibility, safety, and clarity by storing the URL separately.
+
+1. The `live` flag
+    
+    This is being added because we will want to be able to author some 
+    information for programs/courses that don't yet exist on the courseware 
+    platform.
+
+### Loading Course Data
+
+For the time being, the process of reflecting external course data in our
+models can be broken down into 2 parts:
+1. Loading legacy course data
+1. Periodically updating course data from our Open edX instance
+
+##### 1) Loading legacy course data
+
+We will likely receive legacy course data in the form of a data dump. The format 
+for that dump is not yet known. 
+
+**Proposals:**
+1. Create a management command that creates/updates our course data 
+   using either (a) a local file path, or (b) a path to the file in an
+   S3 bucket. Allowing a local file path will make it easier to test and
+   debug.
+
+**Needs Discussion: Data Integrity**
+
+We should not assume that course data will be perfectly loaded in one try using
+one single data dump. We'll probably want to pick one of two approaches:
+
+1. Assume that we will need to load course data from more than one data dump. In
+   this case we should write the command in a way that ensures idempotence. We can
+   discuss the actual implementation if we decide on this approach.
+     - Advantage: Flexibility. We wouldn't need a guarantee that a single data
+       dump was complete and accurate, and we would be able to use the command 
+       after the app hit production.
+     - Disadvantage: More complicated logic.
+1. Assume that a single data dump will contain everything that we need. In this case
+   we should write the command in such a way that it clears all existing data from a
+   given source if you run it more than once (i.e.: if you run the command to load
+   an edx.org data dump after edx.org data has already been loaded, that existing
+   data is deleted first)
+     - Advantage: Easier to implement.
+     - Disadvantage: We could only use this command before the app went live.
+       If something was missed in the first data dump and we needed to amend it somehow, 
+       we would need to do so manually.    
```

Comment:
@rhysyngsun @Ferdi Moving [this discussion](https://github.com/mitodl/mitxpro/pull/23/files#r260794560) to here. 

Given Ferdi's comment that we're expecting to have a complete data dump and load it one time before there is any user activity in the app, it seems that we should go with option [2] here, but it's worth getting more feedback in case I'm missing a scenario where that approach could hurt us

Reply:
I think it's worth clarifying that a single data dump doesn't necessarily correspond to a single import of that data. Specifically:

- It means we pick one of these tradeoffs (my preference for the latter):
  - Import everything in the dump regardless of whether it's needed 
  - Only import what we explicitly need in the short-term and bite off a rework of this if we ever need to pull additional data in
- It also presumes that a one-time import will be 100% correct. Obviously we hope for the best here, but I think we've seen in practice there's often a need for some mix of automated/manual reconciliation of data discrepancies. Again, we can chose to design for this upfront or punt on it until it's an issue (meaning we'd need to fix both this design trade and the actual data issue)

Both these are tradeoffs in terms of initial time/effort vs. being able to deliver data fixes in a timely manner at a later date.

Reply:
Based on what Ferdi said, I think we're making 2 important assumptions:
1. There will be a data dump that has everything we need and nothing that we don't need (e.g.: some extra course runs that we wouldn't want in the new app for some reason)
2. We will not run it after the app goes live and has actual user activity

If those two things are correct, I'm suggesting that we (a) write the command in such a way that rerunning it clears all existing data from the previous import (which we could identify from the `source`), and (b) use Django admin et. al. if any tweaks are needed. Both are low-risk if we aren't going to run this in prod with active users